### PR TITLE
server/api:support offline a store with physically destroyed 

### PR DESCRIPTION
**Pull request by AndreMouche**
Signed-off-by: shirly <AndreMouche@126.com>

### What problem does this PR solve?

It is part of #3076 
This PR support to offline a store and set it physically destroyed which means this store can never up and we can start a new store with the same address 

### What is changed and how it works?

- The API to offline a store
```
DELETE /store/{id}?force=${force}
```
force is a boolean value:
- true: means the store shouldn't be up again and PD can remove it, also we can up a new store with the same address
- false: means the store is offline, it can be up again.

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)

Code changes

- Has HTTP API interfaces change (Don't forget to [add the declarative for API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))

Related changes

https://github.com/pingcap/kvproto/pull/711

### Release note

Offline a store with force true means the store shouldn't be up again and PD can remove it, also we can up a new store with the same address

### CODE REVIEW

File: server/cluster/cluster.go [hunk 1]
```diff
@@ -1021,30 +1021,34 @@ func (c *RaftCluster) RemoveStore(storeID uint64) error {
 	}
 
 	// Remove an offline store should be OK, nothing to do.
-	if store.IsOffline() {
+	if store.IsOffline() && store.IsPhysicallyDestroyed() == physicallyDestroyed {
```
Comment (Yisaer) [inline server/cluster/cluster.go:1024]: Do we need to compare `physicallyDestroyed` here ? i think this is unnecessary.

File: server/cluster/cluster.go [hunk 2]
```diff
@@ -1021,30 +1021,34 @@ func (c *RaftCluster) RemoveStore(storeID uint64) error {
...
 	}
 
 	if store.IsTombstone() {
 		return errs.ErrStoreTombstone.FastGenByArgs(storeID)
 	}
 
-	newStore := store.Clone(core.SetStoreState(metapb.StoreState_Offline))
+	if store.IsPhysicallyDestroyed() {
+		return errors.Errorf("The store is already physically destroyed")
+	}
```
Comment (Yisaer) [inline server/cluster/cluster.go:1034]: Maybe we can return `nil` here as the store is already `PhysicallyDestroyed`

**Review by Yisaer — COMMENTED** [review]
Rest LGTM

File: server/cluster/cluster.go (hunk 2 above)
Comment (AndreMouche) [reply server/cluster/cluster.go:1034]: Since physically-destroyed is not allowed from true to false.

File: server/cluster/cluster.go (hunk 1 above)
Comment (AndreMouche) [reply server/cluster/cluster.go:1024]: we needn't get the lock and set the status again when the status is already as we expected. By the way, it is possible a store may call the API with the following steps:
1. offline the store normally   DELETE store?force=false
2. Find the store is physically destroyed, do not want it to start again: DELETE store?force=true

Comment (Yisaer) [reply server/cluster/cluster.go:1024]: > Find the store is physically destroyed, do not want it to start again: DELETE store?force=true

Why we need to forbid it? I think calling the api `delete` with force=false , then `force=true` is common to see.

File: server/cluster/cluster.go (hunk 2 above)
Comment (Yisaer) [reply server/cluster/cluster.go:1034]: I think we should define a errorCode like `ErrStoreTombstone` and to explain it as `The store have been force removed`

**Review by Yisaer — COMMENTED** [review]
Rest LGTM

File: server/cluster/cluster.go (hunk 1 above)
Comment (HunDunDM) [reply server/cluster/cluster.go:1024]: If there is no judgment here, first `delete && force=false` and then `delete && force=true`, the latter will not be able to mark the `physicallyDestroyed` correctly. I think it needs to compare `physicallyDestroyed`.

File: server/core/store_option.go [hunk 3]
```diff
@@ -82,6 +82,16 @@ func SetStoreState(state metapb.StoreState) StoreCreateOption {
 	}
 }
 
+// OfflineStore offline a store
+func OfflineStore(physicallyDestroyed bool) StoreCreateOption {
```
Comment (rleungx) [inline server/core/store_option.go:86]: Can we just use `func SetStoreState(state metapb.StoreState, Destroyed ...bool) StoreCreateOption` here?

File: server/core/store_option.go [hunk 4]
```diff
@@ -82,6 +82,16 @@ func SetStoreState(state metapb.StoreState) StoreCreateOption {
...
+	return func(store *StoreInfo) {
+		meta := proto.Clone(store.meta).(*metapb.Store)
```
Comment (HunDunDM) [inline server/core/store_option.go:88]: Why is `Clone` needed here? It seems that the `StoreCreateOption ` in this file does not need to clone `meta`.

File: server/core/store_option.go (hunk 3 above)
Comment (HunDunDM) [reply server/core/store_option.go:86]: I think `... bool` is not a good design.

**Review by HunDunDM — APPROVED** [review]

File: server/core/store_option.go (hunk 3 above)
Comment (rleungx) [reply server/core/store_option.go:86]: Or maybe an option?

Comment (HunDunDM) [reply server/core/store_option.go:86]: how about
```

File: server/cluster/cluster.go (hunk 1 above)
Comment (Yisaer) [reply server/cluster/cluster.go:1024]: Thanks for clarifying

**Review by Yisaer — APPROVED** [review]
LGTM

**Pull request merged**